src/ic_layer.cpp

```diff
@@ -0,0 +1,41 @@
+#include <ic_layer.h>
+
+namespace IC {
+    Layer::Layer(const std::string &debugName) : _debugName(debugName) {}
+
+    Layer::~Layer() {}
+
+    LayerStack::LayerStack() {
+        _layerInsertIndex = _layers.begin();
+    }
+
+    LayerStack::~LayerStack() {
+        for (Layer *layer : _layers)
+            delete layer;
+    }
+
+    void LayerStack::PushLayer(Layer *layer) {
+        _layerInsertIndex = _layers.emplace(_layerInsertIndex, layer);
```

Comment:
So the thing about using a iterator here rather than an index into the vector is the iterator _may_ point to invalid data after a modification to the vector occurrs. This is fine here in `PushLayer` because we are recaching the iterator after every modifications, but will become an issue with `PushOverlay` because it does not.
Also, is there a reason to store both Layers and Overlays in a single vector? Otherwise it would be much simpler to split them into two vectors.

Reply:
Updated!